inc/field/fieldsfield.class.php

```diff
@@ -0,0 +1,527 @@
+                                       'display'   => false,
+                                       'condition' => ['is_active' => 1, 'is_deleted' => 0]]);
+            } else {
+               $showuserlink = 0;
+               if (Session::haveRight('user', READ)) {
+                  $showuserlink = 1;
+               }
+               $html.= getUserName($value, $showuserlink);
+            }
+            break;
+         case 'dropdownoperatingsystems':
+            if ($canedit && !$readonly) {
+               $html.= OperatingSystem::dropdown(['name'      => $this->field->fields['name'],
+                                       'value'     => $value,
+                                       'entity'    => -1,
+                                       'right'     => 'all',
+                                       'display'   => false//,
+                                       /*'condition' => 'is_active=1 && is_deleted=0'*/]);
+            } else {
+               $os = new OperatingSystem();
+               $os->getFromDB($value);
+               $html.= $os->fields['name'];
+            }
+      }
+
+      unset($_SESSION['plugin']['fields']['values_sent']);
+      return $html;
+   }
+
+   public function serializeValue(): string {
+      return $this->value;
+   }
+
+   public function getRawValue() {
+      return $this->value;
+   }
+
+   public function deserializeValue($value) {
+      $this->value = $value;
+   }
+
+   public function getValueForDesign(): string {
+      return "";
+   }
+
+   public function getValueForTargetText($domain, $richText): ?string {
+      return "";
+   }
+
+   public function getValueForApi(): string {
+      return "";
+   }
+
+   public function isValidValue($value): bool {
+
+      if (!is_null($this->field)) {
+
+         //check data type for input number / url
+         $valid = true;
+         if ($this->field->fields['type'] == 'number' && !empty($this->value) && !is_numeric($this->value)) {
+            $number_errors[] = $this->field->fields['label'];
+            $valid = false;
+         } else if ($this->field->fields['type'] == 'url' && !empty($this->value)) {
+            if (filter_var($this->value, FILTER_VALIDATE_URL) === false) {
+               $url_errors[] = $this->field->fields['label'];
+               $valid = false;
+            }
+         }
+
+         if (!empty($number_errors)) {
+            Session::AddMessageAfterRedirect(__("Some numeric fields contains non numeric values", "fields").
+                                             " : ".implode(', ', $number_errors), false, ERROR);
+         }
+
+         if (!empty($url_errors)) {
+            Session::AddMessageAfterRedirect(__("Some URL fields contains invalid links", "fields").
+                                             " : ".implode(', ', $url_errors), false, ERROR);
+         }
+
+         if (!$valid) {
+            return false;
+         }
+
+         //All is OK
+         return true;
+      }
```

Comment:
If you don't enter the if statement then the method won't return any value. I suggest the following, with lowering code complexity. The if now  exits early,  then I inverted the condition.

```suggestion
      if (is_null($this->field)) {
         return false;
      }

      //check data type for input number / url
      $valid = true;
      if ($this->field->fields['type'] == 'number' && !empty($this->value) && !is_numeric($this->value)) {
         $number_errors[] = $this->field->fields['label'];
         $valid = false;
      } else if ($this->field->fields['type'] == 'url' && !empty($this->value)) {
         if (filter_var($this->value, FILTER_VALIDATE_URL) === false) {
            $url_errors[] = $this->field->fields['label'];
            $valid = false;
         }
      }

      if (!empty($number_errors)) {
         Session::AddMessageAfterRedirect(__("Some numeric fields contains non numeric values", "fields").
                                          " : ".implode(', ', $number_errors), false, ERROR);
      }

      if (!empty($url_errors)) {
         Session::AddMessageAfterRedirect(__("Some URL fields contains invalid links", "fields").
                                          " : ".implode(', ', $url_errors), false, ERROR);
      }

      if (!$valid) {
         return false;
      }

      //All is OK
      return true;

      }
```